python/ray/data/impl/compute.py

```diff
@@ -193,10 +201,7 @@ def map_block_nosplit(
                 list(tasks.keys()), timeout=0.01, num_returns=1, fetch_local=False
             )
             if not ready:
-                if (
-                    len(workers) < self.max_size
-                    and len(ready_workers) / len(workers) > 0.8
-                ):
+                if len(workers) < max_size and len(ready_workers) / len(workers) > 0.8:
```

Comment:
```suggestion
                if len(workers) < max_size and len(ready_workers) / len(workers) > 0.8 and blocks_in:
```

@clarkzinzow how about this instead? I think this is actually better, since it will give a tighter dynamic bound to the pool size if say there are only a few tasks left.

We could also kill off workers below if a worker is ready but there are no blocks left to process.

Reply:
@ericl Good call out! I opted to do both with a slight tweak:
1. Bound pool size by the initial number of input blocks, adjusted for the max in-flight tasks per worker. This is required for at least the `min_size`, otherwise we might create too many workers [upfront](https://github.com/ray-project/ray/blob/7cd4b40e6f12c1a26ee2a1815f0eb22752bf2402/python/ray/data/impl/compute.py#L194).
2. Bound pool size by the remaining work to do. On each autoscaling loop, we know that `starting = len(workers) - len(ready_workers)` workers are still initializing and that `len(blocks_in)` contains our remaining work, so we shouldn't allocate a new worker if we already have enough workers initializing to maximally parallelize the remaining work. This is also adjusted for the max in-flight tasks per worker, where we want at most `len(blocks_in) // self.max_tasks_in_flight_per_actor` workers to transform the remaining blocks, so we only start a new worker if `starting` is less than that.

Downscaling logic was straightforward: on every finished worker creation task or transformation task, terminate the worker if (a) there's no work left to do, and (b) there are no tasks in-flight on the worker.